datasync/downloader/downloader.go

```diff
@@ -577,6 +584,108 @@ func (d *Downloader) spawnSync(fetchers []func() error, peerID string) error {
 	return err
 }
 
+func (d *Downloader) SyncStakingInfo(id string, from, to uint64) error {
+	if d.fixing == true {
+		return errors.New("already syncing")
+	}
+	logger.Info("start syncing staking infos", "from", from, "to", to)
+	d.fixing = true
+
+	var hashes []common.Hash
+	from = params.CalcStakingBlockNumber(from)
+	for i := from; i <= to; i += params.StakingUpdateInterval() {
+		if hash, has, err := reward.HasStakingInfoFromDB(i); err == nil && !has {
+			if hash == (common.Hash{}) {
+				d.fixing = false
+				return fmt.Errorf("failed to retrieve block hash by number (blockNumber: %v)", i)
+			}
+			d.fixingBlocks = append(d.fixingBlocks, i)
+			hashes = append(hashes, hash)
+		}
+	}
+
+	if len(d.fixingBlocks) == 0 && len(hashes) == 0 {
+		d.fixing = false
+		return fmt.Errorf("there is no staking info to be synced")
+	}
+
+	conn := d.peers.Peer(id)
+	if conn == nil {
+		d.fixing = false
+		return errors.New("the given peer is not registered")
+	}
+
+	d.fixingTotal = len(d.fixingBlocks)
+
+	go func() {
+		defer func() {
+			d.fixing = false
+			d.fixingBlocks = []uint64{}
+			d.fixingTotal = 0
+		}()
+		d.fixingCh = make(chan []*reward.StakingInfo, 1)
+
+		fixed := 0
+		for {
+			timer := time.NewTimer(30 * time.Second)
+			if len(hashes) > MaxStakingInfoFetch {
+				go conn.peer.RequestStakingInfo(hashes[:MaxStakingInfoFetch])
+				logger.Info("requested staking infos", "num", MaxStakingInfoFetch)
+				hashes = hashes[MaxStakingInfoFetch:]
+			} else if len(hashes) > 0 {
+				go conn.peer.RequestStakingInfo(hashes)
+				logger.Info("requested staking infos", "num", len(hashes))
+				hashes = []common.Hash{}
+			} else {
+				logger.Error("no more requests, but not completed", "not completed", len(d.fixingBlocks))
+				return
+			}
+
+			select {
+			case <-timer.C:
+				logger.Error("timeout")
+				return
+			case stakingInfos := <-d.fixingCh:
+				for _, stakingInfo := range stakingInfos {
+					if d.fixingBlocks[0] != stakingInfo.BlockNum {
+						logger.Error("failed to receive expected block", "expected", d.fixingBlocks[0], "actual", stakingInfo.BlockNum)
+						return
+					}
+
+					if err := reward.AddStakingInfoToDB(stakingInfo); err != nil {
+						logger.Error("failed to add staking info", "fixed", fixed, "stakingInfo", stakingInfo, "err", err)
+						return
+					}
+					fixed++
+					d.fixingBlocks = d.fixingBlocks[1:]
+				}
+
+				if len(d.fixingBlocks) == 0 {
+					logger.Info("syncing staking info is finished", "fixed", fixed)
+					return
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+type SyncingStatus struct {
+	Syncing   bool     `json:"syncing"`
+	Pending   []uint64 `json:"pending"`
+	Total     int      `json:"total"`
+	Completed int      `json:"completed"`
+}
+
+func (d *Downloader) SyncStakingInfoStatus() *SyncingStatus {
+	return &SyncingStatus{
```

Comment:
Is it thread safe to request SyncStakingInfoStatus when synching?

Reply:
I think it would be fine since it only reads variables. Do you think that it is necessary to lock the variables?

Reply:
@jeongkyun-oh Yes, I think the variables can be mismatched each other.

Reply:
I think it doesn't matter because it provides an approximate progress in real time. You can ignore this comment.